third_party/move/move-compiler-v2/src/env_pipeline/lambda_lifter.rs

```diff
@@ -185,13 +187,358 @@ impl<'a> LambdaLifter<'a> {
+        }
+
+        // Add captured LocalVar parameters
+        for (name, var_info) in mem::take(&mut self.free_locals) {
+            let var_info_id = var_info.node_id;
+            let ty = env.get_node_type(var_info_id);
+            let loc = env.get_node_loc(var_info_id);
+            if var_info.modified {
+                env.error(
+                    &loc,
+                    &format!(
+                        "captured variable `{}` cannot be modified inside of a lambda", // TODO(LAMBDA)
+                        name.display(env.symbol_pool())
+                    ),
+                );
+            }
+            params.push(Parameter(name, ty.clone(), loc.clone()));
+            let new_id = env.new_node(loc, ty);
+            if let Some(inst) = env.get_node_instantiation_opt(var_info_id) {
+                env.set_node_instantiation(new_id, inst);
+            }
+            closure_args.push(ExpData::LocalVar(new_id, name).into_exp())
+        }
+
+        (params, closure_args, param_index_mapping)
+    }
+
+    fn get_arg_if_simple(arg: &Exp) -> Option<&Exp> {
+        use ExpData::*;
+        match &arg.as_ref() {
+            Value(..) | LocalVar(..) | Temporary(..) => Some(arg),
+            Sequence(_, exp_vec) => {
+                if let [exp] = &exp_vec[..] {
+                    Self::get_arg_if_simple(exp)
+                } else {
+                    None
+                }
+            },
+            Invalid(..) | Call(..) | Invoke(..) | Lambda(..) | Quant(..) | Block(..)
+            | IfElse(..) | Match(..) | Return(..) | Loop(..) | LoopCont(..) | Assign(..)
+            | Mutate(..) | SpecBlock(..) => None,
+        }
+    }
+
+    // If final `args` match `lambda_params`, and all other args are simple, then returns
+    // the simple prefix of `args`.
+    fn get_args_if_simple<'b>(
+        lambda_params: &Vec<Parameter>,
+        args: &'b [Exp],
+    ) -> Option<Vec<&'b Exp>> {
+        if lambda_params.len() <= args.len() {
+            let mut simple_args: Vec<&Exp> = args
+                .iter()
+                .filter_map(|exp| Self::get_arg_if_simple(exp))
+                .collect();
+            if simple_args.len() == args.len()
+                && lambda_params
+                    .iter()
+                    .rev()
+                    .zip(simple_args.iter().rev())
+                    .all(|(param, arg)| {
+                        if let ExpData::LocalVar(_, name) = arg.as_ref() {
+                            *name == param.get_name()
+                        } else {
+                            false
+                        }
+                    })
+            {
+                let remaining_size = args.len() - lambda_params.len();
+                simple_args.truncate(remaining_size);
+                return Some(simple_args);
+            }
+        }
+        None
+    }
+
+    // Only allow simple expressions which cannot vary or abort
```

Comment:
What is an expression that can "vary"?

Reply:
A function call.  I also don't want things to be too expensive, in case pulling them out of the lambda changes semantics.  I changed the comment to indicate that.

If we weren't in rust, then I might be worried about a value that has references taken, but reference semantics should take care of that: